osm2streets/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,229 @@
 ///
 /// The existing complexity field is ignored, so be careful how you use the guessed value.
-fn guess_complexity(streets: &StreetNetwork, intersection_id: &NodeID) -> IntersectionComplexity {
-    let roads = streets.roads_per_intersection(*intersection_id);
+pub fn guess_complexity(
+    streets: &StreetNetwork,
+    intersection_id: &NodeID,
+) -> (IntersectionComplexity, ConflictType, Vec<Movement>) {
+    use ConflictType::*;
+    let roads: Vec<_> = streets
+        .roads_per_intersection(*intersection_id)
+        .iter()
+        .map(|id| &streets.roads[id])
+        .filter(|road| road.is_driveable())
+        .collect();
 
     // A terminus is characterised by a single connected road.
     if roads.len() == 1 {
-        return Terminus;
+        return (Terminus, Uncontested, Vec::new());
     }
 
-    // A Connection is characterised by exactly two connected roads.
-    if roads.len() == 2 {
-        return Connection;
-    }
+    // Calculate all the possible movements, (except U-turns, for now).
+    let mut connections = Vec::new();
+    // Consider all pairs of roads, from s to d. Identify them using their index in the list - which
+    // is sorted in clockwise order - so that we can compare their position later.
+    for s in 0..roads.len() {
+        for d in 0..roads.len() {
+            if s == d {
+                continue; // Ignore U-turns.
+            }
 
-    // A MultiConnection is characterised by exactly one dividing line traveling through it (the
-    // line that separates traffic in different directions) and no traffic that crosses it.
-    if roads.len() == 3 {
-        let mut _num_roads_in = 0;
-        let mut _num_roads_out = 0;
-        let mut num_roads_inout = 0;
-        for road_id in roads {
-            let is_outward = road_id.i1 == *intersection_id;
-            let road = streets.roads.get(&road_id).unwrap();
-            match road.oneway_for_driving() {
-                Some(Direction::Fwd) => {
-                    if is_outward {
-                        _num_roads_out += 1
-                    } else {
-                        _num_roads_in += 1
-                    }
-                }
+            // Calculate if it is possible to emerge from s into the intersection.
+            let src_road = roads[s];
+            if !can_drive_out_of(src_road, *intersection_id) {
+                continue;
+            }
 
-                Some(Direction::Back) => {
-                    if is_outward {
-                        _num_roads_in += 1
-                    } else {
-                        _num_roads_out += 1
-                    }
-                }
-                None => num_roads_inout += 1,
+            // Calculate if it is possible to leave the intersection into d.
+            let dst_road = roads[d];
+            if !can_drive_into(dst_road, *intersection_id) {
+                continue;
+            }
+
+            // TODO detect U-Turns that should be assumed forbidden.
```

Comment:
These conditions sound necessary + sufficient to me! Happy to leave this for a followup PR